### PR TITLE
fix(upgrade): strengthen version upgrade validation semantics

### LINKED ISSUE #2183: Strengthen upgrade version validation semantics in validateVersionToUpgrade
### Is your feature request related to a problem?

The current upgrade validation logic in `validateVersionToUpgrade` appears to primarily validate minor-version differences without performing explicit semantic-version compatibility checks.

As a result, unsupported upgrade paths and ambiguous semver edge cases may unintentionally pass validation. Examples include:
- Unsupported major-version upgrades (e.g. `1.x -> 2.x`)
- Downgrade attempts
- Same-version upgrades
- Invalid semantic version inputs
- Potential multi-minor upgrade edge cases

This weakens upgrade safety guarantees and may lead to undefined behavior during cluster or component upgrade workflows.

Since upgrade validation is a critical reliability and compatibility boundary, stricter semantic validation would improve upgrade safety, predictability, and maintainability.

### Describe the solution you'd like

I would like `validateVersionToUpgrade` to implement stricter and more explicit semantic-version validation rules.

Suggested improvements:
- Explicitly validate semantic version parsing for both source and target versions
- Reject invalid semver inputs early with clear errors
- Reject downgrade attempts
- Reject same-version upgrade attempts
- Add explicit major-version compatibility validation
- Allow only supported upgrade paths (for example: adjacent minor-version upgrades if that is the intended policy)

Additionally, test coverage should be expanded using table-driven tests to cover important upgrade scenarios and edge cases, including:
- downgrade attempts
- same-version upgrades
- valid one-minor upgrades
- unsupported major-version upgrades
- invalid semver inputs
- unsupported multi-minor upgrades (if applicable)

The implementation should remain minimal, idiomatic, and aligned with the project's existing upgrade policy and semver conventions.

### How would users interact with this feature?

Users would interact with this indirectly through existing upgrade workflows.

Example scenarios:
- Cluster upgrade validation
- Component upgrade orchestration
- CLI/API-driven upgrade operations

The validation layer would prevent unsupported upgrade paths before upgrade execution begins, providing earlier and clearer feedback to users and operators.

Example expected behavior:
Allowed:
- `1.4.0 -> 1.5.0`

Rejected:
- `1.4.0 -> 1.4.0`
- `1.4.0 -> 1.3.0`
- `1.4.0 -> 2.0.0`
- invalid semver strings

### Alternative solutions considered

Alternative approaches considered:

1. Keeping the current minor-version-only validation
   - This does not sufficiently protect against unsupported major-version upgrades or malformed semver input.

2. Allowing unrestricted semver upgrades
   - This may simplify validation logic but introduces operational and compatibility risks.

3. Deferring validation entirely to downstream upgrade logic
   - This delays error detection and weakens safety guarantees.

A dedicated semantic-version validation layer provides clearer upgrade semantics and earlier failure detection.

### Additional Context

This improvement would strengthen upgrade reliability and reduce ambiguity around supported upgrade paths.

Since upgrade validation is a high-impact control point, stronger semver semantics and expanded edge-case testing would improve long-term maintainability and operational safety.

Suggested implementation area:
- `upgrade.go`
- `validateVersionToUpgrade`

Suggested testing approach:
- idiomatic Go table-driven tests
- reuse existing semver utilities/libraries already present in the repository

### PR DESCRIPTION
**Pull request by Harishrs2006**
Summary:
Strengthened upgrade version validation semantics in `validateVersionToUpgrade`.

Changes:
* Added explicit major-version validation
* Rejected downgrade and same-version upgrades
* Improved invalid semver handling
* Expanded edge-case test coverage

Tests:
Added tests for:
* downgrade attempts
* same-version upgrades
* valid one-minor upgrades
* unsupported major-version upgrades
* invalid semver inputs

Validated locally with:

```bash id="jlwm92"
go test ./pkg/cli/upgrade -v
```

No breaking changes. Existing valid upgrade paths remain supported.

Fixes: #2183

### CODE REVIEW

File: CODE_CHANGES_SUMMARY.md [hunk 1]
```diff
@@ -0,0 +1,316 @@
+# Code Changes Summary
```
Comment (Denyme24) [inline CODE_CHANGES_SUMMARY.md]: please refrain from pushing these irrelevant files.

**Review by Denyme24 — COMMENTED** [review]
discard all those irrelevant md files.

**Comment by Harishrs2006** [comment]
@Denyme24 
Removed the unrelated markdown files — the PR now only contains changes to pkg/cli/upgrade/upgrade.go and pkg/cli/upgrade/upgrade_test.go. Sorry for the noise!
